### PR TITLE
Support mapping properties in XML and YAML

### PR DESCRIPTION
**Pull request by meyerbaptiste**
| Q | A |
| --- | --- |
| Bug fix? | no |
| New feature? | yes |
| BC breaks? | no |
| Deprecations? | no |
| Tests pass? | yes |
| Fixed tickets | #739, api-platform/api-platform#140 |
| License | MIT |
| Doc PR | N/A |

This PR adds support of mapping for properties in XML and YAML.
I followed the recommendations from #421.

### CODE REVIEW

File: src/Metadata/Property/Factory/XmlPropertyMetadataFactory.php [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+     * @param \SimpleXMLElement $element
+     *
+     * @return array
+     */
+    private function getAttributes(\SimpleXMLElement $element) : array
+    {
+        $attributes = [];
+        foreach ($element->attribute as $attribute) {
+            $value = isset($attribute->attribute[0]) ? $this->getAttributes($attribute) : (string) $attribute;
+            isset($attribute['name']) ? $attributes[(string) $attribute['name']] = $value : $attributes[] = $value;
```
Comment (soyuka) [inline src/Metadata/Property/Factory/XmlPropertyMetadataFactory.php]: I find this hard to read, I'm not found of ternary value attributions, maybe it's just me

File: src/Metadata/Property/Factory/YamlPropertyMetadataFactory.php [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+    private function getMetadata(string $resourceClass, string $property) : array
+    {
+        foreach ($this->paths as $path) {
+            try {
+                $resources = Yaml::parse(file_get_contents($path));
+            } catch (ParseException $parseException) {
+                $parseException->setParsedFile($path);
+
+                throw $parseException;
+            }
```
Comment (soyuka) [inline src/Metadata/Property/Factory/YamlPropertyMetadataFactory.php]: nice one, maybe we should add this to the other Yaml factories!

File: src/Metadata/Property/Factory/XmlPropertyMetadataFactory.php (hunk 1 above)
Comment (teohhanhui) [inline src/Metadata/Property/Factory/XmlPropertyMetadataFactory.php]: I find this hard to read too. I think an if-else is just fine here.

File: src/Metadata/Property/Factory/XmlPropertyMetadataFactory.php [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+
+            $properties = (new \DOMXPath($domDocument))->query(sprintf('//resources/resource[@class="%s"]/property[@name="%s"]', $resourceClass, $propertyName));
+
+            if (false === $properties || 0 >= $properties->length || null === $properties->item(0) || false === $property = simplexml_import_dom($properties->item(0))) {
+                continue;
+            }
+
+            return [
+                'description' => (string) $property['description'] ?: null,
+                'readable' => 'true' === (string) $property['readable'] || 'false' === (string) $property['readable'] ? constant((string) $property['readable']) : null,
```
Comment (dunglas) [inline src/Metadata/Property/Factory/XmlPropertyMetadataFactory.php]: Why not using [`XmlUtils::phpize`](https://github.com/symfony/symfony/blob/master/src/Symfony/Component/Config/Util/XmlUtils.php#L184) instead of this trick?

File: src/Metadata/Property/Factory/XmlPropertyMetadataFactory.php [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+     * Creates a new instance of metadata if the property is not already set.
+     *
+     * @param PropertyMetadata $propertyMetadata
+     * @param array            $metadata
+     *
+     * @return PropertyMetadata
+     */
+    private function update(PropertyMetadata $propertyMetadata, array $metadata) : PropertyMetadata
+    {
+        foreach (['description' => 'get', 'readable' => 'is', 'writable' => 'is', 'writableLink' => 'is', 'readableLink' => 'is', 'required' => 'is', 'identifier' => 'is', 'iri' => 'get', 'attributes' => 'get'] as $metadataKey => $accessorPrefix) {
```
Comment (teohhanhui) [inline src/Metadata/Property/Factory/XmlPropertyMetadataFactory.php]: Probably split the array into multiple lines.

File: src/Metadata/Resource/Factory/XmlResourceMetadataFactory.php [hunk 5]
```diff
@@ -83,9 +83,9 @@ private function getMetadata(string $resourceClass) : array
                 }
 
                 return [
-                    (string) $resource['shortName'] ?? null,
-                    (string) $resource['description'] ?? null,
-                    (string) $resource['iri'] ?? null,
+                    (string) $resource['shortName'] ?: null,
```
Comment (dunglas) [inline src/Metadata/Resource/Factory/XmlResourceMetadataFactory.php]: Why this change? What if the `shortName` key doesn't exist?

File: src/Metadata/Property/Factory/XmlPropertyMetadataFactory.php (hunk 3 above)
Comment (dunglas) [inline src/Metadata/Property/Factory/XmlPropertyMetadataFactory.php]: Btw,  `(string) $property['readable']` can be `'0'` or `'1'` too according to the XSD spec.

File: src/Metadata/Resource/Factory/XmlResourceMetadataFactory.php (hunk 5 above)
Comment (soyuka) [inline src/Metadata/Resource/Factory/XmlResourceMetadataFactory.php]: +1, anyway I thought this was tested but I've the feeling it is not and should be

Comment (meyerbaptiste) [inline src/Metadata/Resource/Factory/XmlResourceMetadataFactory.php]: Correct me if I am wrong but if an attribute does not exist in a `\SimpleXMLElement` class, its value is `null` so the result of this expression is never `null` (i.e. `(string) null` => `''`).

Example with `$resource['foo']` which does not exist:
```php
var_dump($resource); // class SimpleXMLElement
var_dump($resource['foo']); // NULL
var_dump((string) $resource['foo'] ?? null); // string(0) ""
var_dump((string) $resource['foo'] ?: null); // NULL
```

Comment (dunglas) [inline src/Metadata/Resource/Factory/XmlResourceMetadataFactory.php]: I wasn't aware of this behavior, nice catch!

File: src/Metadata/Property/Factory/XmlPropertyMetadataFactory.php (hunk 3 above)
Comment (meyerbaptiste) [inline src/Metadata/Property/Factory/XmlPropertyMetadataFactory.php]: What about this?
```php
in_array((string) $property['foo'], ['true', 'false', '0', '1'], true) ? filter_var((string) $property['foo'], FILTER_VALIDATE_BOOLEAN) : null;
```

**Comment by meyerbaptiste** [comment]
Updated with your changes @soyuka @dunglas @teohhanhui.

Travis failed due to depreciation notice from Twig (see twigphp/Twig#2208). This will be fixed by the next release of `symfony/twig-bundle` (i.e. `3.1.6`).

File: src/Metadata/Property/Factory/YamlPropertyMetadataFactory.php (hunk 2 above)
Comment (meyerbaptiste) [inline src/Metadata/Property/Factory/YamlPropertyMetadataFactory.php]: Done :)

**Comment by dunglas** [comment]
@meyerbaptiste can you use this trick https://github.com/twigphp/Twig/issues/2208#issuecomment-256316598 to make Travis happy please?

**Comment by meyerbaptiste** [comment]
I tested it yesterday, Travis failed again with `--prefer-lowest`.

**Comment by dunglas** [comment]
You probably need to bump the minimal Twig version in `composer.json` to make Travis happy with `--prefer-lowest`.

**Comment by meyerbaptiste** [comment]
It's okay @dunglas, `3.1.6` is now released.

File: src/Metadata/Property/Factory/XmlPropertyMetadataFactory.php [hunk 6]
```diff
@@ -0,0 +1,194 @@
...
+            } catch (PropertyNotFoundException $propertyNotFoundException) {
+                // Ignore not found exception from decorated factories
+            }
+        }
+
+        if (!class_exists($resourceClass) || !property_exists($resourceClass, $property) || empty($propertyMetadata = $this->getMetadata($resourceClass, $property))) {
+            return $this->handleNotFound($parentPropertyMetadata, $resourceClass, $property);
+        }
+
+        if ($parentPropertyMetadata) {
```
Comment (Simperfit) [inline src/Metadata/Property/Factory/XmlPropertyMetadataFactory.php]: `null !== $parentPropertyMetadata`

File: src/Metadata/Property/Factory/XmlPropertyMetadataFactory.php [hunk 7]
```diff
@@ -0,0 +1,194 @@
...
+     * @param string                $resourceClass
+     * @param string                $property
+     *
+     * @throws PropertyNotFoundException
+     *
+     * @return PropertyMetadata
+     */
+    private function handleNotFound(PropertyMetadata $parentPropertyMetadata = null, string $resourceClass, string $property) : PropertyMetadata
+    {
+        if ($parentPropertyMetadata) {
```
Comment (Simperfit) [inline src/Metadata/Property/Factory/XmlPropertyMetadataFactory.php]: `null !== $parentPropertyMetadata`

File: src/Metadata/Property/Factory/XmlPropertyMetadataFactory.php [hunk 8]
```diff
@@ -0,0 +1,194 @@
...
+        foreach ($this->paths as $path) {
+            try {
+                $domDocument = XmlUtils::loadFile($path, self::RESOURCE_SCHEMA);
+            } catch (\InvalidArgumentException $e) {
+                throw new InvalidArgumentException($e->getMessage(), $e->getCode(), $e);
+            }
+
+            $properties = (new \DOMXPath($domDocument))->query(sprintf('//resources/resource[@class="%s"]/property[@name="%s"]', $resourceClass, $propertyName));
+
+            if (false === $properties || 0 >= $properties->length || null === $properties->item(0) || false === $property = simplexml_import_dom($properties->item(0))) {
```
Comment (Simperfit) [inline src/Metadata/Property/Factory/XmlPropertyMetadataFactory.php]: Could you change this for the sake of lisibility ?
`if ( blabla
   || 
     blabla)`

File: src/Metadata/Property/Factory/XmlPropertyMetadataFactory.php (hunk 6 above)
Comment (meyerbaptiste) [inline src/Metadata/Property/Factory/XmlPropertyMetadataFactory.php]: Yes but https://github.com/api-platform/core/pull/770#discussion_r80776281.

File: src/Metadata/Property/Factory/XmlPropertyMetadataFactory.php (hunk 7 above)
Comment (meyerbaptiste) [inline src/Metadata/Property/Factory/XmlPropertyMetadataFactory.php]: Same here.

File: src/Metadata/Property/Factory/XmlPropertyMetadataFactory.php (hunk 8 above)
Comment (meyerbaptiste) [inline src/Metadata/Property/Factory/XmlPropertyMetadataFactory.php]: Done @Simperfit.

**Comment by dunglas** [comment]
Thanks @meyerbaptiste. Don't forget to update the docs please.
Btw, it would be nice to have snippet of JS on the website hiding alternatives to annotations by default like on the Symfony's website.
